File: backend/app/middleware/security_headers.py
"""Security response headers for the API (defense-in-depth).

The browser-facing clickjacking/XSS surface is the SPA (served by the frontend
container, which sets its own page-level CSP — see frontend/securityHeaders.mjs).
These headers harden the JSON API responses too. The API serves JSON only, so a
maximally-strict CSP (`default-src 'none'`) is appropriate — nothing should ever be
loaded from an API origin.

HSTS is opt-in (``settings.enable_hsts``) so it is never emitted over plain HTTP and
is safe to enable only where TLS terminates in front of the app (e.g. a reverse proxy).
"""

from __future__ import annotations

from fastapi import Request, Response

from ..core.config import settings

# Applied to every API response. setdefault() is used at call time so a route that
# deliberately set one of these (rare) is never clobbered.
_STATIC_HEADERS = {
    "X-Content-Type-Options": "nosniff",
    "X-Frame-Options": "DENY",
    "Content-Security-Policy": "default-src 'none'; frame-ancestors 'none'",
    "Referrer-Policy": "no-referrer",
    "Cross-Origin-Opener-Policy": "same-origin",
    "Cross-Origin-Resource-Policy": "same-site",
}


async def security_headers_middleware(request: Request, call_next):
    """Set hardening headers on every response (registered outermost in main.py)."""
    response: Response = await call_next(request)
    for name, value in _STATIC_HEADERS.items():
        response.headers.setdefault(name, value)
    if settings.enable_hsts:
        response.headers.setdefault(
            "Strict-Transport-Security", "max-age=63072000; includeSubDomains"
        )
    return response

File: frontend/securityHeaders.mjs
// Security response headers for the SPA shell (served by server.mjs). Applied to the
// HTML/static responses only — the /api proxy forwards the backend's own headers.
//
// The CSP is enforcing, widened only where the app genuinely needs it:
//   - 'wasm-unsafe-eval'      IGV compiles WebAssembly (e.g. CRAM decoding)
//   - img-src/worker-src blob: IGV/canvas render via blob URLs and web workers
//   - connect-src https:       presigned-S3 host (config-driven) + remote IGV tracks
//   - Google Fonts            style/font from fonts.googleapis.com / fonts.gstatic.com
//   - frame-ancestors 'none'   blocks clickjacking (no view is meant to be embedded)
// If an unforeseen origin is blocked in staging, switch the header name to
// `Content-Security-Policy-Report-Only` to observe before enforcing.
export const CSP = [
  "default-src 'self'",
  "script-src 'self' 'wasm-unsafe-eval'",
  "style-src 'self' 'unsafe-inline' https://fonts.googleapis.com",
  "font-src 'self' https://fonts.gstatic.com data:",
  "img-src 'self' data: blob:",
  "connect-src 'self' https:",
  "worker-src 'self' blob:",
  "frame-ancestors 'none'",
  "object-src 'none'",
  "base-uri 'self'",
  "form-action 'self'",
].join('; ');

/**
 * Express middleware that sets the SPA security headers. HSTS is read at call time
 * (opt-in via ENABLE_HSTS) so it is only emitted behind a TLS-terminating proxy.
 */
export function securityHeaders(_req, res, next) {
  res.setHeader('Content-Security-Policy', CSP);
  res.setHeader('X-Content-Type-Options', 'nosniff');
  res.setHeader('X-Frame-Options', 'DENY');
  res.setHeader('Referrer-Policy', 'no-referrer');
  res.setHeader('Cross-Origin-Opener-Policy', 'same-origin');
  if (process.env.ENABLE_HSTS === 'true') {
    res.setHeader('Strict-Transport-Security', 'max-age=63072000; includeSubDomains');
  }
  next();
}
